lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -1,17 +1,1035 @@
+"""Tests for the Tool Meister module.
 """
-Tests for the Tool Meister modules.
-"""
+
+from http import HTTPStatus
 import io
 import logging
-import shutil
-import subprocess
-from http import HTTPStatus
 from pathlib import Path
-
 import pytest
 import responses
+import shutil
+import signal
+import subprocess
+from typing import Any, List, NamedTuple, Tuple
+
+from pbench.agent.tool_meister import (
+    log_raw_io_output,
+    Tool,
+    TransientTool,
+    PcpTransientTool,
+    PersistentTool,
+    DcgmTool,
+    NodeExporterTool,
+    PcpTool,
+    ToolException,
+    ToolMeister,
+    ToolMeisterError,
+)
+
+
+def test_log_raw_io_output(caplog):
+    class MockedPipe:
+        def readlines(self, hint: int = -1) -> bytes:
+            yield b"line1\n"
+            yield b"\n"
+            yield b"line3\n"
+            yield b"line4\n"
+
+    log_raw_io_output(MockedPipe(), logging.getLogger("test_log_raw_io_output"))
+
+    assert caplog.record_tuples[0][2] == "line1"
+    assert caplog.record_tuples[1][2] == "line3"
+    assert caplog.record_tuples[2][2] == "line4"
+
+
+class NullObject:
+    pass
+
+
+class MockShutilWhich:
+    """Mock shutil.which in the tool_meister module."""
+
+    def __init__(self, find: List[str]):
+        """Constructor
+
+        Arguments:
+
+            find:  list of things that should be found
+        """
+        self.find = frozenset(find)
+
+    def which(self, exec: str) -> str:
+        """Mocked version of shutil.which.
+
+        Arguments:
+
+            exec:  A string representing the executable to "look for"
+
+        Returns a "/usr/bin/{exec}" if the exectuable is in the list of things
+        to be found, otherwise None.
+        """
+        return f"/usr/bin/{exec}" if exec in self.find else None
+
+
+def mock_shutil_which(monkeypatch, find: List[str] = []):
+    """Encapsulate steps to mock out shutil.which.
+
+    Arguments:
+
+        monkeypatch:  the monkeypatch instance to use
+        find:  list of things that should be found
+    """
+    msw = MockShutilWhich(find)
+
+    monkeypatch.setattr(shutil, "which", msw.which)
```

Comment:
So, I did [a little reading](https://docs.python.org/3/reference/datamodel.html#index-35) and a bunch of experimenting, and I think I now have a rudimentary understanding of Python function values and monkeypatching them.

The key is that there are two Python types, `<class 'function'>` and `<class 'method'>`, and they are both callable, but they behave differently with respect to the parameter list.  "Function objects" include normal Python functions and `@staticmethod` class member functions.  "(Instance) Method objects" include `@classmethod` class member functions.  The kicker is that normal class member functions can produce **_either_** function objects _or_ method objects, _depending on how you get them_ -- if you use an **instance** to access the function value, you get a method object, whereas, if you use the class, you get a function object:
```
>>> class MyClass:
...   def my_member_func(*args, **kwargs):
...     print(f"In MyClass.my_member_func, args = {args}, kwargs = {kwargs}")
... 
>>> func_ref = MyClass.my_member_func
>>> type(func_ref)
<class 'function'>
>>> method_ref = MyClass().my_member_func
>>> type(method_ref)
<class 'method'>
```

If you make a call using a method object, then a reference to the instance (either the class, for a class method, or the object instance, for an instance method) gets inserted at the beginning of the argument list, and it's bound to the first formal parameter.  If you make a call using a function object, then only the arguments from the call are bound to the parameters.
```
>>> method_ref(1, 2, 3, last=4)
In MyClass.my_member_func, args = (<__main__.MyClass object at 0x10ed13310>, 1, 2, 3), kwargs = {'last': 4}
>>> func_ref(1, 2, 3, last=4)
In MyClass.my_member_func, args = (1, 2, 3), kwargs = {'last': 4}
```

So, in the code which Peter has here, the mock for `which` is invoked with `self` bound to `msw`, because the monkeypatch is done with an _instance method object_.  In the mock which @npalaska was trying to use for `Path`, the `Path` object was getting "lost" because he was using an instance method object for a class which was _not_ a `Path`, and so the mock object was bound to `self` instead of the desired `Path` object.  When he changed the code to use a _function object_ for the patch, his mock function received all the arguments, including the `Path`, because the he changed the member function _in the class_ which was then invoked on an _instance_, and so it called his code with the instance reference as the first parameter.

So, contrary to my earlier assertions here and elsewhere, we don't _have_ to make sure that mocks for member functions are themselves member functions (although, this is probably still a good idea, but it will probably only work if we mock the whole object and not just the one function).  Instead, we have to make damn sure that we understand whether the value we hand to `monkeypatch` is a _function object_ or a _method object_ as well as whether it will be reached via a method call on an _instance_ or not.  😝

Reply:
Tracked in #2725.